src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -0,0 +1,1349 @@
+    // Return Value:
+    // - <none>
+    void ControlCore::Search(const winrt::hstring& text,
+                             const bool goForward,
+                             const bool caseSensitive)
+    {
+        if (text.size() == 0)
+        {
+            return;
+        }
+
+        const Search::Direction direction = goForward ?
+                                                Search::Direction::Forward :
+                                                Search::Direction::Backward;
+
+        const Search::Sensitivity sensitivity = caseSensitive ?
+                                                    Search::Sensitivity::CaseSensitive :
+                                                    Search::Sensitivity::CaseInsensitive;
+
+        ::Search search(*GetUiaData(), text.c_str(), direction, sensitivity);
+        auto lock = _terminal->LockForWriting();
+        if (search.FindNext())
+        {
+            _terminal->SetBlockSelection(false);
+            search.Select();
+            _renderer->TriggerSelection();
+        }
+    }
+
+    void ControlCore::SetBackgroundOpacity(const float opacity)
+    {
+        if (_renderEngine)
+        {
+            _renderEngine->SetDefaultTextBackgroundOpacity(::base::saturated_cast<float>(opacity));
+        }
+    }
+
+    // Method Description:
+    // - Asynchronously close our connection. The Connection will likely wait
+    //   until the attached process terminates before Close returns. If that's
+    //   the case, we don't want to block the UI thread waiting on that process
+    //   handle.
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - <none>
+    winrt::fire_and_forget ControlCore::_asyncCloseConnection()
+    {
+        if (auto localConnection{ std::exchange(_connection, nullptr) })
+        {
+            // Close the connection on the background thread.
+            co_await winrt::resume_background();
+            localConnection.Close();
+            // connection is destroyed.
+        }
+    }
+
+    void ControlCore::Close()
+    {
+        if (!_closing.exchange(true))
+        {
+            // Stop accepting new output and state changes before we disconnect everything.
+            _connection.TerminalOutput(_connectionOutputEventToken);
+            _connectionStateChangedRevoker.revoke();
+
+            // GH#1996 - Close the connection asynchronously on a background
+            // thread.
+            // Since TermControl::Close is only ever triggered by the UI, we
+            // don't really care to wait for the connection to be completely
+            // closed. We can just do it whenever.
+            _asyncCloseConnection();
+
+            {
+                // GH#8734:
+                // We lock the terminal here to make sure it isn't still being
+                // used in the connection thread before we destroy the renderer.
+                // However, we must unlock it again prior to triggering the

```

Comment:
this lock scares me w/ the async connection teardown

Reply:
however, we already disconnected the events.

Reply:
I mean, this whole function has always ben scary, but at least this bit is verbatim from `main`